build: create merge script w/ cherry-picking

devversion opened the pull request:
Creates a merge script that automatically merges pull
requests from Github and cherry-picks them into the
right publish branches based on the PR target label.

The script has been designed in an extensible and programmatically
acessible way, so that it can be used in other Angular repositories too.

The script provides infrastructure for PR merge safety checks. e.g.
it currently enforces that the CLA is signed, or that the merge ready
label is assigned, or that no CI checks are failing / pending.

devversion commented on scripts/merge-script/config.ts:
```diff
@@ -0,0 +1,74 @@
+    errors.push('No merge ready label configured.');
+  }
+  return errors;
+}
```
The config parsing will be easy to integrate w` `getDevInfraConfig()` if we eventually move it to `/dev-infra`.

josephperrott commented on scripts/merge-script/cli.ts:
```diff
@@ -0,0 +1,183 @@
+  const githubToken = cliGithubToken || process.env.GITHUB_TOKEN || process.env.TOKEN;
+
+  if (!githubToken) {
+    console.error(
```
I wonder if it would be worth creating a common util that contains `error()`, `info()` and `log()`, that perform `console.error`, `console.info`, etc. with the `chalk` colors already applied?   We are going to have to do a ton of logging things like this throughout our tooling and consistency would be nice.

If thats something that sounds like a good idea to you, I would also throw out the suggestion of these functions taking in multiple args, and placing each arg on its own console line.

It would allow us to have this block become, IMO, more readable as
```js
if (!githubToken) {
  logging.error(
    'No Github token is set. Please set the `GITHUB_TOKEN` environment variable.',
    'Alternatively, pass the `--github-token` command line flag.');
  process.exit(1);
}
```

devversion replied:
I like the idea of having a shared logging util. The only thing that concerns me is that usually the chalk statements are needed in the middle of a message. e.g `chalk.red(<..> ${chalk.bold(...)} <..>)`. That might be inconsistent then with the mid-sentence color changes.

Regardless of that though, I think that it would make sense to explore this. Potentially we could also have a `debug` method then (that is based on an environment variable).

Should we do that separately in the `dev-infra/` package?

josephperrott replied:
Yeah, I think its probably worth waiting and just doing it in the dev-infra package.

josephperrott commented on scripts/merge-script/cli.ts:
```diff
@@ -0,0 +1,183 @@
+        failures.every(failure => FORCE_OVERRIDABLE_FAILURES.includes(failure.status)) :
+        false;
+
+    if (status === MergeStatus.SUCCESS) {
```
I personally find this more readable, with `if` blocks that have returns since we are also either exiting the function or the entire process on each match.

devversion replied:
I converted it to a `switch` statement. Let me know if that is better.

devversion commented on merge-config.js:
```diff
@@ -0,0 +1,41 @@
+module.exports = () => {
```
Yeah, we most likely will. I was thinking about making it support a JS file when this is moved over. Do you have any concerns with that?

devversion commented on scripts/merge-script/cli.ts:
```diff
@@ -0,0 +1,183 @@
+ */
+async function main() {
+  const {config, pullRequests, force, dryRun, githubToken: cliGithubToken} = parseCommandLine();
+  const githubToken = cliGithubToken || process.env.GITHUB_TOKEN || process.env.TOKEN;
```
Yep. Changed!

josephperrott commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,333 @@
+    } catch {
+      return {prNumber, status: PullRequestStatus.UNKNOWN_MERGE_ERROR};
+    } finally {
+      // Always try to restore the branch if possible. We don't want to leave the
```
Additional to always trying to restore the branch we came from, we should also clean up all of the branches we created correct?

devversion commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,333 @@
+      }
+      // The state is set to `pending` if a pull request does not have any status checks.
+      // https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-ref
+      if (state === 'pending' && statuses.length !== 0 && !force) {
```
That's a good point. Done.

devversion commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,333 @@
+    }
+
+    for (const pullRequest of queue) {
+      const failure = await this._mergePullRequest(pullRequest);
```
Yeah. Done!

mmalerba commented on scripts/merge-script/cli.ts:
```diff
@@ -0,0 +1,135 @@
+}
+
+/** Parses the command line and returns the passed options. */
+function parseCommandLine(): {
```
Should we use some kind of command line parsing library? I'm always tempted to write stuff like this in Python since they have really nice parsing built in

devversion replied:
Yeah agreed! The plan is to use `yargs` for this if we move it over to `dev-infra/` in `angular/angular`. We decided to do that as a follow-up.

jelbourn replied:
Add a TODO?

jelbourn commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,243 @@
+   * @returns A pull request failure or null in case of success.
+   */
+  private async _mergePullRequest(
+      {prNumber, targetBranches, requiredBaseSha, commitsAmount}: PullRequest,
```
Could this destructuring be done in the method instead of in the definition? I just think it looks cleaner that way.

devversion replied:
Yeah, that's a good point.

jelbourn commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,243 @@
+   * merge as it causes unnecessary fetch requests.
+   * @returns A pull request failure or null in case of success.
+   */
+  private async _mergePullRequest(
```
It's not immediately clear what the difference between `merge` and `_mergePullRequest` is from the name. Maybe use names like `mergeAll` and `mergeSingle`?

devversion replied:
Agreed. I'll try to make this more obvious.

jelbourn commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,243 @@
+      {prNumber, targetBranches, requiredBaseSha, commitsAmount}: PullRequest,
+      dryRun: boolean): Promise<PullRequestFailure|null> {
+    let previousBranch: string|null = null;
+    try {
```
Add a comment about which error(s) you're expecting with this `try`

jelbourn commented on scripts/merge-script/commit-message-filter.js:
```diff
@@ -0,0 +1,38 @@
+  const lines = message.split(/\n/);
+  // Add the pull request number to the commit message title. This matches what
+  // Github does when PRs are merged on the web through the `Squash and Merge` button.
+  lines[0] += ' (#' + prNumber + ')';
```
Use template strings?

devversion replied:
Yep. I forgot that when I copied the script from `angular/angular`.

jelbourn commented on scripts/merge-script/config.ts:
```diff
@@ -0,0 +1,74 @@
+
+/** Reads and validates the configuration file at the specified location. */
+export function readAndValidateConfig(filePath: string): {config?: Config, errors?: string[]} {
+  try {
```
Add comment about the type of error(s) you're expecting with this `try`

jelbourn commented on scripts/merge-script/console.ts:
```diff
@@ -0,0 +1,55 @@
+    case PullRequestStatus.NOT_MERGE_READY:
+      return `Not marked as merge ready.`;
+    case PullRequestStatus.CHERRY_PICK_CONFLICTS:
+      return `Could not cherry-pick pull request into branch: ${context}`;
```
Probably want to use the phrase "merge conflict" in this message; the current wording omits the reason the change can't be cherry-picked

jelbourn commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,243 @@
+      dryRun: boolean): Promise<PullRequestFailure|null> {
+    let previousBranch: string|null = null;
+    try {
+      previousBranch = this._git(['rev-parse', '--abbrev-ref', 'HEAD']).stdout.trim();
```
I'd add a comment like "Remember the branch we're on now so we can change back to it at the end"

devversion commented:
@jelbourn Addressed your feedback (and also simplified things a bit). Also, are you aware of the fact that this script doesn't preserve the purple `Merged` state for pull requests? I investigated whether we can merge purely with the Github API, but that's not possible. Additionally, we can't merge using the API, and then cherry-pick into other branches as that would mean that autosquashing doesn't work. Autosquashing is commonly used in framework.

I think we should just acknowledge that this is a Github limitation. Github should just provide an API for changing the PR state to `merged`. Clearly merges using the `Squash & Merge` button aren't fast-forward either.. so Github is already tricking here.

jelbourn commented:
Ah, if we can't preserve the purple "merged" status, then I would prefer to continue pressing the button in the UI. Maybe then the script could handle the cherry-picking after merging the PR.

devversion commented:
@jelbourn I think the problem here is that this script should be also usable in other repositories. If it would be only used in the components repository, then we could definitely use the Github PR merge API, and perform the cherry-picking locally. 

Though I wanted the script to be part of the shared dev-infra tools, so that it can be used in framework too. Framework relies on auto-squashing which isn't supported by the Github merge buttons. We don't perform fast-forward merges (we cannot), so if we ever want that behavior, Github needs to provide an API for us to set the merged state.

Alternatively we can add opt-in options to the script that allow:

*  Squash merging through pressing the button in the UI. The script only does the cherry-picking. This would also simplify the process of doing manual commit message cleanup.
* Squash merging through the API (automatically). The script does the cherry-picking as needed.

Basically, the script would be made "suitable" for requirements from both repositories where the script is used. I personally would prefer having a single canonical way, but I understand the reason for having it show up as `Merged`. At the same time though.. I'd like to start using `fixup!` commits in the components repo, and not having to request for preserving of multiple commits (if needed).

Cherry-picking a merged PR through the Github API/UI also will complicate things, as it will be difficult for the script to find corresponding PR commits if merged through the `Rebase and Merge` button.

devversion commented:
@jelbourn @josephperrott I made the changes we discussed in our meeting about this. The merge strategy for using the Github API turned a bit more code than initially anticipated, but I think that amount of code is reasonable and acceptable. Easy to remove if we ever want a canonical merge strategy across all repositories.

josephperrott commented on scripts/merge-script/config.ts:
```diff
@@ -0,0 +1,102 @@
+   * The downside is that fixup or squash commits no longer work as the Github API does
+   * not support this.
+   */
+  githubApiMerge?: {
+    default: GithubApiMergeMethod,
+    labels?: {pattern: string, method: GithubApiMergeMethod}[],
+  };
```
Can we require this is defined so its explicit which is being chosen?

```suggestion
  githubApiMerge: false | {
    default: GithubApiMergeMethod,
    labels?: {pattern: string, method: GithubApiMergeMethod}[],
  };
```

devversion replied:
Yep. Done

josephperrott commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  /** Gets whether the current Git repository has uncommitted changes. */
+  private _hasUncommittedChanges(): boolean {
```
Does it make more sense for this to be a method on `GitClient`?

devversion replied:
Yeah, I think that makes sense. Done.

josephperrott commented on scripts/merge-script/index.ts:
```diff
@@ -0,0 +1,120 @@
+      return {status: MergeStatus.DIRTY_WORKING_DIR};
+    }
+
+    const pullRequestOrFailure = await loadAndValidatePullRequest(this, prNumber, force);
+
+    if (!isPullRequest(pullRequestOrFailure)) {
+      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
+    }
+
+    const pullRequest: PullRequest = pullRequestOrFailure;
```
Is it possible for this to be done with a type guard?

`isPullRequest` can confirm the type of `pullRequestOrFailure`?

That would allow this to be:

```suggestion
    const pullRequest = await loadAndValidatePullRequest(this, prNumber, force);

    if (!isPullRequest(pullRequest)) {
      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
    }
```

devversion replied:
`isPullRequest` already is a type guard. The only difference I see here is that you renamed the variable from `pullRequestOrFailure` to `pullRequest`. I'm fine with that.

josephperrott commented on scripts/merge-script/pull-request.ts:
```diff
@@ -0,0 +1,128 @@
+  targetBranches: string[];
+  /** Branch that the PR targets in the Github UI. */
+  githubTargetBranch: string;
+  /** Amount of commits this pull request contains. */
```
nit, can we do this as `commitCount`?

devversion replied:
Haha 😂 You always get me with that one. Sure, will change it!

josephperrott commented on scripts/merge-script/pull-request.ts:
```diff
@@ -0,0 +1,128 @@
+   * Optional context for the pull request failure. e.g. Name of the branch that
+   * the pull request could not be cherry-picked into.
+   */
+  context?: any;
```
I think to do so, we would need to overload this interface for each of the potential `PullRequestStatus`es, I am not entirely should how much value it would add.  But could definitely be considered.

devversion replied:
Yeah, I agree that this is not really ideal. My goal here is that the actual merge logic does not have to deal with writing messages (also since that avoids duplication).

I'm changing it to an abstract class with static methods on it. e.g.

```ts
return PullRequestFailure.mergeConflicts(failedBranches)
```

Let me know if that works for you all.

josephperrott commented on scripts/merge-script/strategies/api-merge.ts:
```diff
@@ -0,0 +1,191 @@
+      return {status: PullRequestStatus.MISMATCHING_TARGET_BRANCH, context: targetBranches};
+    }
+
+    // In case a required base is specified for this pull request, check if the pull
```
`// In cases where a required base`?

josephperrott commented on scripts/merge-script/strategies/api-merge.ts:
```diff
@@ -0,0 +1,191 @@
+    // In case a required base is specified for this pull request, check if the pull
+    // request contains the given commit. If not, return a pull request failure. This
+    // check is useful for enforcing that PRs are rebased on top of a given commit. e.g.
+    // a commit that changes the codeowner ship validation. PRs which are not rebased
```
s/codeowner ship/code ownership

josephperrott commented on scripts/merge-script/strategies/api-merge.ts:
```diff
@@ -0,0 +1,191 @@
+    // then run into cherry-pick conflicts after the initial merge already completed.
+    const failure = await this._checkMergability(pullRequest, cherryPickTargetBranches);
+
+    // If the PR could not be cherry-picked into all  locally, then we know it can't be done
```
`into all target branches locally`

josephperrott commented on scripts/merge-script/strategies/api-merge.ts:
```diff
@@ -0,0 +1,191 @@
+    // If a commit message fixup is needed and we are merging using squash, we prompt
+    // for commit message changes. Note that this does not work with other merge methods
+    // as the Github API only allows commit message modifications for squash merging.
+    if (needsCommitMessageFixup && method === 'squash') {
```
Should we error in cases in which something is marked as `needsCommitMessageFixup` but is a `method` that is not `squash`?

devversion replied:
I like that.

devversion replied:
done.

josephperrott commented on scripts/merge-script/strategies/api-merge.ts:
```diff
@@ -0,0 +1,191 @@
+      return {status: PullRequestStatus.MERGE_CONFLICTS, context: failedBranches};
+    }
+
+    this.pushTargetBranchesUpstream(cherryPickTargetBranches);
```
Should we confirm the pushes are successful?  What happens if it were to fail here?

devversion replied:
If the push fails, then the script exits and the failure is printed out (git stderr). I don't think there is much we can do. This probably can happen if multiple PRs are merged at the same time. In those cases, the Git stderr message should explain this, and it's up to the consumer of this script to re-run (if possible).

josephperrott commented on scripts/merge-script/strategies/api-merge.ts:
```diff
@@ -0,0 +1,191 @@
+      message: 'Please update the commit message',
+      default: commitMessage,
+    });
+
```
Should we also be validating the commit message that is created here?

devversion replied:
I'd say yes, but that seems to be something we can tackle as a follow-up? Especially since we are not having commit message validation in the components repo yet. I'd like to set it up soon, but I think we need some upstream changes to account for a different `scope` that we support.

josephperrott replied:
SGTM

josephperrott commented on scripts/merge-script/strategies/strategy.ts:
```diff
@@ -0,0 +1,127 @@
+import {PullRequest, PullRequestFailure} from '../pull-request';
+
+/** Name of a temporary branch that contains the head of a currently-processed PR. */
+export const TEMP_PR_HEAD_BRANCH = 'merge_pr_head';
```
Should this branch name be unique?  Or is it okay that it gets reused?

devversion replied:
It can be any arbitrary branch name. This is the one from the framework's merge script.

devversion replied:
Tried to make this more clear in the description.

devversion commented on scripts/merge-script/cli.ts:
```diff
@@ -0,0 +1,138 @@
+
+  // Merge the pull request and handle the result. If failures have been reported, exit
+  // the process with a non-zero exit code.
+  if (!await handleMergeResult(await api.merge(prNumber, false), force)) {
```
👍 Done.

devversion commented on scripts/merge-script/cli.ts:
```diff
@@ -0,0 +1,138 @@
+function parseCommandLine():
+    {config: Config, force: boolean, prNumber: number, dryRun?: boolean, githubToken: string} {
+  const {config: configPath, githubToken: githubTokenArg, force, _: [prNumber]} =
+      minimist<any>(process.argv.slice(2), {
```
Unfortunately not. We'll clean this up anyway in a follow-up when we switch to `yargs`.

devversion commented:
@jelbourn @josephperrott Addressed most of the feedback. Please have another look.

josephperrott commented on scripts/merge-script/pull-request.ts:
```diff
@@ -0,0 +1,104 @@
+    return PullRequestFailure.claUnsigned();
+  }
+
+  const {data: {state}} =
+      await git.api.repos.getCombinedStatusForRef({...git.repoParams, ref: prData.head.sha});
+
+  const targetLabel = getTargetLabelFromPullRequest(config, labels);
+  if (targetLabel === null) {
+    return PullRequestFailure.noTargetLabel();
+  }
```
Can we move the data request to after the target label check?  Just so we don't request the data and then fail before we even use it.
```suggestion
  const targetLabel = getTargetLabelFromPullRequest(config, labels);
  if (targetLabel === null) {
    return PullRequestFailure.noTargetLabel();
  }
  
  const {data: {state}} =
      await git.api.repos.getCombinedStatusForRef({...git.repoParams, ref: prData.head.sha});
```

devversion replied:
Yeah, fixed!
